Content.Server/Database/ServerDbManager.cs

```diff
@@ -86,131 +48,360 @@ public void Init()
                     var options = CreateSqliteOptions();
-                    _db = new ServerDbSqlite(options);
-                    break;
+                    return new SqliteServerDbContext(options);
                 case "postgres":
                     options = CreatePostgresOptions();
-                    _db = new ServerDbPostgres(options);
-                    break;
+                    return new PostgresServerDbContext(options);
                 default:
                     throw new InvalidDataException("Unknown database engine {engine}.");
             }
         }
 
-        public Task<PlayerPreferences> InitPrefsAsync(NetUserId userId, ICharacterProfile defaultProfile)
+        public async Task<PlayerPreferences> InitPrefsAsync(NetUserId userId, ICharacterProfile defaultProfile)
         {
-            return _db.InitPrefsAsync(userId, defaultProfile);
+
+            var profile = new Profile
+            {
+                Slot = 0
+            };
+            ((HumanoidCharacterProfile) defaultProfile).ConvertProfile(profile);
+
+            var prefs = await ServerDbContext.Preferences
+                .Where(p => p.UserId == userId)
+                .SingleOrDefaultAsync();
+
+            // To explain: The code later will return null if it encounters preferences and no profiles.
+            // This is because the preferences are corrupt, and the DB code doesn't have access to the function to create a new profile.
+            // In this case, a Preference object already exists but is safe to reuse.
+            if (prefs is null)
+            {
+                prefs = new Preference
+                {
+                    UserId = userId.UserId
+                };
+                ServerDbContext.Preferences.Add(prefs);
+                Logger.WarningS("c.s.db.serverdbmanager", "Reinitializing preferences[{0}] (user {1}) - this only happens in erroneous situations.", prefs.Id, userId);
+            }
+
+            ServerDbContext.PreferenceProfiles
+                .Add(new PreferenceProfile {
+                    Preference = prefs,
+                    Profile = profile
+                });
+
+            prefs.Profiles = new List<Profile> { profile };
+
+            await ServerDbContext.SaveChangesAsync();
+
+            return new PlayerPreferences(new[] {new KeyValuePair<int, ICharacterProfile>(0, defaultProfile)}, 0);
         }
 
-        public Task SaveSelectedCharacterIndexAsync(NetUserId userId, int index)
+        public async Task SaveSelectedCharacterIndexAsync(NetUserId userId, int index)
         {
-            return _db.SaveSelectedCharacterIndexAsync(userId, index);
+            var preference = await ServerDbContext.Preferences
+                .Where(p => p.UserId == userId)
+                .SingleAsync();
+
+            var selected = await ServerDbContext.PreferenceProfiles
+                .Where(p => p.PreferenceId == preference.Id)
+                .SingleOrDefaultAsync();
```

Comment:
These can be done in one query with a join. Though I would prefer if this entire `PreferenceProfiles` thing was removed in the first place since it should just be a two-way relationship...